server/instance_cloud.go

```diff
@@ -63,6 +63,28 @@ func newCloudInstance(p *Plugin, key types.ID, installed bool, rawASC string, as
 	}
 }
 
+func (p *Plugin) installCloudInstance(rawURL string) (string, *serverInstance, error) {
+	jiraURL, err := utils.CheckJiraURL(p.GetSiteURL(), rawURL, false)
+	if err != nil {
+		return "", nil, err
+	}
+	if !utils.IsJiraCloudURL(jiraURL) {
+		return "", nil, errors.Errorf("`%s` is a Jira server URL, not a Jira Cloud", jiraURL)
+	}
+
+	instance := &serverInstance{
```

Comment:
Since we are using a `serverInstance`, we are using logic for OAuth 1. I recommend we use Jira Cloud's OAuth 2.0 option instead: https://developer.atlassian.com/cloud/jira/platform/oauth-2-3lo-apps. We do miss an opportunity to reuse code from the server instance code, but the benefits of implementing OAuth 2.0 are well worth it.

One reason this is beneficial is because we can [publish the app](https://developer.atlassian.com/cloud/jira/platform/oauth-2-3lo-apps/#distributing-your-oauth-2-0--3lo--apps) on Jira's marketplace, making it a one-click install for admins.

Note that this requires a consistent redirect URL to be used on our end. We have an existing mechanism to support a single redirect URL for all Mattermost Cloud instances https://github.com/mattermost/chimera. On-prem customers will need to manually set up the OAuth app in Jira, though this is something already expected for their case. Here's an [example](https://github.com/mattermost/mattermost-plugin-github/blob/e516e139823fa3bfaaed59433fc588119a5aa470/server/plugin/plugin.go#L351) of the GitHub plugin using the Chimera proxy.

We also have an existing pattern to support OAuth 2.0 functionality in the plugin projects defined [here](https://github.com/mattermost/mattermost-plugin-api/blob/73008cfda2fbefe305f704eeb45975fabe39d7f2/experimental/oauther/oauth2.go#L36). Here's a minimal [example](https://github.com/mickmister/mattermost-plugin-oauth-example/pull/1/files) I made showing how it can be used. We will need to be able to support multiple Jira instances with this, and I think the API for the `OAuther` is flexible enough to support this. It requires very little code/logic on the plugin side if we use the `OAuther` tool.

Reply:
Great! Let's use OAuth2! At this moment I wanted to test if with a few lines of code I could replicate the same functionality over Jira Cloud. I'll check out the Chimera examples and the examples.

Thank you!

Reply:
Supporting Chimera is probably out of scope of this PR. It's certainly a "nice to have" that we want to support, but it requires more testing etc. so we can leave it out of the initial implementation.